processor/transformprocessor/internal/metrics/func_extract_sum_metric.go

```diff
@@ -0,0 +1,119 @@
+}
+
+func createExtractSumMetricFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	args, ok := oArgs.(*extractSumMetricArguments)
+
+	if !ok {
+		return nil, fmt.Errorf("extractSumMetricFactory args must be of type *extractSumMetricArguments")
+	}
+
+	return extractSumMetric(args.AggTemp, args.Monotonic)
+}
+
+// this interface helps unify the logic for extracting data from different histogram types
+// all supported metric types' datapoints implement it
+type SumCountDataPoint interface {
+	Attributes() pcommon.Map
+	Sum() float64
+	Count() uint64
+	StartTimestamp() pcommon.Timestamp
+	Timestamp() pcommon.Timestamp
+}
+
+func extractSumMetric(aggTempEnum ottl.Enum, monotonic bool) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	aggTemp := pmetric.AggregationTemporality(aggTempEnum)
+	switch aggTemp {
+	case pmetric.AggregationTemporalityDelta, pmetric.AggregationTemporalityCumulative:
+	default:
+		return nil, fmt.Errorf("unknown aggregation temporality: %s", aggTemp.String())
+	}
+
+	return func(_ context.Context, tCtx ottlmetric.TransformContext) (interface{}, error) {
+		metric := tCtx.GetMetric()
+
+		switch metric.Type() {
+		case pmetric.MetricTypeHistogram, pmetric.MetricTypeExponentialHistogram, pmetric.MetricTypeSummary:
+		default:
+			return nil, nil
+		}
+
+		sumMetric := createSumMetric(
+			tCtx.GetMetrics(),
+			metric.Name()+"_sum",
+			metric.Description(),
+			metric.Unit(),
+			aggTemp,
+			monotonic,
+		)
+
+		switch metric.Type() {
+		case pmetric.MetricTypeHistogram:
+			dataPoints := metric.Histogram().DataPoints()
+			for i := 0; i < dataPoints.Len(); i++ {
+				addSumDataPoint(dataPoints.At(i), sumMetric.Sum().DataPoints())
+			}
+		case pmetric.MetricTypeExponentialHistogram:
+			dataPoints := metric.ExponentialHistogram().DataPoints()
+			for i := 0; i < dataPoints.Len(); i++ {
+				addSumDataPoint(dataPoints.At(i), sumMetric.Sum().DataPoints())
+			}
+		case pmetric.MetricTypeSummary:
+			dataPoints := metric.Summary().DataPoints()
+			for i := 0; i < dataPoints.Len(); i++ {
+				addSumDataPoint(dataPoints.At(i), sumMetric.Sum().DataPoints())
+			}
+		default:
+			return nil, nil
+		}
+
+		return nil, nil
+	}, nil
+}
+
+func createSumMetric(
+	metrics pmetric.MetricSlice,
+	name string,
+	description string,
+	unit string,
+	aggregationTemporality pmetric.AggregationTemporality,
+	isMonotonic bool,
+) pmetric.Metric {
+	sumMetric := metrics.AppendEmpty()
+	sumMetric.SetDescription(description)
+	sumMetric.SetName(name)
+	sumMetric.SetUnit(unit)
+	sumMetric.SetEmptySum().SetAggregationTemporality(aggregationTemporality)
+	sumMetric.Sum().SetIsMonotonic(isMonotonic)
+	return sumMetric
+}
+
+func addSumDataPoint(dataPoint SumCountDataPoint, destination pmetric.NumberDataPointSlice) {
+	newDp := destination.AppendEmpty()
+	dataPoint.Attributes().CopyTo(newDp.Attributes())
+	newDp.SetDoubleValue(dataPoint.Sum())
```

Comment:
The [`sum` field is optional](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L419), so we need to do a `HasSum` check here or we will either cause a panic or produce invalid data.

I'm not sure what to do if the datapoint has no sum, but it seems like we may have to just ignore the error or log out that we hit this case. Otherwise we could throw an error here, but would need a way of checking for "datapoints within a histogram metric all have sums", which seems oddly specific.

Reply:
Huh, I was confident it was required, just like `count`, good catch. I think I agree that ignoring datapoints without sums is the best choice here; as a user, that would be the behaviour I'd expect and find most useful. Returning an error for the whole metric may be more correct in some sense, but we'd require that the user manually check for this, which is not that easy as you point out.